src/prefect/engine/result/base.py

```diff
@@ -123,6 +126,18 @@ def store_safe_value(self) -> None:
                 value=value, result_handler=self.result_handler
             )
 
+    def serialize(self) -> str:
+        return base64.b64encode(cloudpickle.dumps(self.value)).decode()
+
+    @classmethod
+    def deserialize(cls, serialized_value: str) -> Any:
+        return cloudpickle.loads(base64.b64decode(serialized_value))
+
+    def render_filepath(self) -> str:
+        if not self.filepath_template:
+            raise ValueError("No filepath_template provided")
+        return self.filepath_template.format(**prefect.context)
```

Comment:
GREETINGS talked with @cicdw more on what was being discussed on slack at https://prefecthq.slack.com/archives/CH84XHRB4/p1584460170003700, we want to make this more like

```
def format(self, **kwargs: Any) -> Result:
     # something like...
     self._templated_filename = self.filename_template.format(**kwargs)
     return self
```

At this stage, this is basically so that Result object itself isn't so implicitly dependent on prefect.context, which could vary depending when the rendering method is called; it's more explicit that we get to decide in the pipeline logic when we send it context which is also when we know what is in context better.

For example, later on the pipeline is going to do something like this, so its really clear its just moving the **prefect.context expansion to where the pipeline has control:

```
hydrated_result = self.task.result.format(**prefect.context)
hydrated_result.write(val)
state = Success(result=hydrated_result)
return state
```

In addition, to Chris' concerns on the slack thread it makes it easier for us to send the rendered filename to Cloud since it's in an attribute of Result. 
On the one hand this is 'nice to have' so that Cloud UI can render the filename and users can easily inspect that from the UI.
On another hand Chris was explaining that Cloud will need to know this because of the way it sets up flow recovery. In particular the edge case he is thinking of that resonated the most with me is:
  - Cloud user sets up a task with `Result.filename_template="results/{task_name}.txt"`
      - In the suggested situation, when the task writes its result, it also sets its Result.templated_filename = "/Users/laura/.prefect/results/my_super_task.json"
   -  task runs successfully and submits its state to Cloud, including that templated_filename
   -  drama: right when the task finished running my agent and/or infra dies, but there are still downstream dependencies in this flow run
    -  when I wake up my agent/infra, because I am a Cloud user my flow is going to try and start itself again from Cloud's database which has all the task states
        - Cloud grabs the task states from its database to initialize a FlowRunner to recover the flow run we were on, including deserializing upstream results to pass to the next unrun task
        - if upstream results have Result.templated_filename set already, it can call Result.read to grab that result from that location and give it to the next task
       - if the upstream results only have Result.render_filepath and try to do so with the current `prefect.context`, they are actually in FlowRunner context instead of TaskRunner context and might not be able to figure it out if the template requires TaskRunner context like {task_name}.